### ISSUE TITLE
Loss of connectivity after time: unable to mallocAndStrcpy_s

### ISSUE THREAD

**Issue by rtheil-growlink** [Azure/azure-iot-sdk-c#2271]
Working on ESP32 devices here and having a new issue where suddenly devices will disconnect for no apparent reason and never recover.

In the following log, you can see that I receive a confirmation that a message was received by Azure. Immediately after there's an error "unable to mallocAndStrcpy_s option value". After this occurs the last 5 lines of the log will repeat over and over until I reset the controller. It's never able to recover.

I've gotten detailed info about the heap and there's typically at least 40k contiguous space free, with over 70k total free. This doesn't seem to be a memory issue despite the error, unless I'm missing something. Also worth noting is that this error started popping up only after the update of the SDK to v1.9 from v1.4

```
-> 07:58:01 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_LEAST_ONCE | TOPIC_NAME: devices/000000000000d6b8/messages/events/ | PACKET_ID: 2210 | PAYLOAD_LEN: 124
<- 07:58:01 PUBACK | PACKET_ID: 2210
I (25958846) AzureService: Confirmation[2178] received for message tracking id = 2178 with result = IOTHUB_CLIENT_CONFIRMATION_OK
Error: Time:Wed Apr  6 07:58:10 2022 File:../lib/azure-iot-sdk-c/c-utility/pal/tlsio_options.c Func:tlsio_options_clone_option Line:119 unable to mallocAndStrcpy_s option value
Error: Time:Wed Apr  6 07:58:10 2022 File:../lib/azure-iot-sdk-c/c-utility/pal/tlsio_options.c Func:local_clone_option Line:231 Unexpected local_clone_option failure
Error: Time:Wed Apr  6 07:58:10 2022 File:../lib/azure-iot-sdk-c/c-utility/src/optionhandler.c Func:AddOptionInternal Line:76 unable to clone value
Error: Time:Wed Apr  6 07:58:10 2022 File:../lib/azure-iot-sdk-c/c-utility/src/optionhandler.c Func:OptionHandler_Clone Line:186 Error cloning option TrustedCerts
Error: Time:Wed Apr  6 07:58:10 2022 File:../lib/azure-iot-sdk-c/c-utility/src/optionhandler.c Func:AddOptionInternal Line:76 unable to clone value
Error: Time:Wed Apr  6 07:58:10 2022 File:../lib/azure-iot-sdk-c/c-utility/src/xio.c Func:xio_retrieveoptions Line:287 unable to OptionHandler_AddOption
-> 07:58:10 DISCONNECT
W (25968556) AzureService: Connection Status result:IOTHUB_CLIENT_CONNECTION_UNAUTHENTICATED, Connection Status reason: IOTHUB_CLIENT_CONNECTION_EXPIRED_SAS_TOKEN
E (25972986) esp-tls-mbedtls: No server verification option set in esp_tls_cfg_t structure. Check esp_tls API reference
E (25972986) esp-tls-mbedtls: Failed to set client configurations
E (25972996) esp-tls: create_ssl_handle failed
Error: Time:Wed Apr  6 07:58:46 2022 File:../lib/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:UpdateMqttConnectionStateIfNeeded Line:2783 mqtt_client timed out waiting for CONNACK
Error: Time:Wed Apr  6 07:58:46 2022 File:../port/src/tlsio_esp_tls.c Func:tlsio_esp_tls_destroy Line:153 tlsio_esp_tls_destroy called while not in TLSIO_STATE_CLOSED.
E (26004006) esp-tls-mbedtls: No server verification option set in esp_tls_cfg_t structure. Check esp_tls API reference
E (26004006) esp-tls-mbedtls: Failed to set client configurations
E (26004016) esp-tls: create_ssl_handle failed
```

Any ideas why this might be happening?

**Comment by danewalton** [comment]
Hi @rtheil-growlink 
[Can you replace the code here](https://github.com/Azure/azure-c-shared-utility/blob/a0915d48fcd491416ac7e6d89abba4e98cdc2382/pal/tlsio_options.c#L116-L121) with this snippet below so we can see what the return value is from `mallocAndStrcpy_s()`?

```c
        *out_value = NULL;
        int res;
        if ((res = mallocAndStrcpy_s((char**)out_value, value)) != 0)
        {
            LogError("unable to mallocAndStrcpy_s option value: %d", res);
            result = TLSIO_OPTIONS_RESULT_ERROR;
        }
```

**Comment by rtheil-growlink** [comment]
I haven't had this happen since last week. I will close this for now and reopen when it comes back up since that code snippet is in place.

**Comment by rtheil-growlink** [comment]
Finally got the controller to throw this error again. Happened twice over the weekend.

```
Error: Time:Mon Apr 18 01:46:03 2022 File:../lib/azure-iot-sdk-c/c-utility/pal/tlsio_options.c Func:tlsio_options_clone_option Line:120 unable to mallocAndStrcpy_s option value: 12
Error: Time:Mon Apr 18 01:46:03 2022 File:../lib/azure-iot-sdk-c/c-utility/pal/tlsio_options.c Func:local_clone_option Line:232 Unexpected local_clone_option failure
Error: Time:Mon Apr 18 01:46:03 2022 File:../lib/azure-iot-sdk-c/c-utility/src/optionhandler.c Func:AddOptionInternal Line:76 unable to clone value
Error: Time:Mon Apr 18 01:46:03 2022 File:../lib/azure-iot-sdk-c/c-utility/pal/tlsio_options.c Func:tlsio_options_retrieve_options_ex Line:275 unable to save TrustedCerts option
Error: Time:Mon Apr 18 01:46:03 2022 File:../lib/azure-iot-sdk-c/c-utility/src/xio.c Func:xio_retrieveoptions Line:279 unable to concrete_io_retrieveoptions
```

**Comment by momuno** [comment]
result 12 is ENOMEM.  malloc is returning NULL.
```
        char* temp = (char*)malloc(l + 1);


        /*Codes_SRS_CRT_ABSTRACTIONS_99_037: [Upon failure to allocate memory for the destination, the function will return ENOMEM.]*/
        if (temp == NULL)
        {
            result = ENOMEM;
        }
```
https://github.com/Azure/azure-c-shared-utility/blob/7ddf017f1c08a82688baf08cbe6111ae38e297b8/src/crt_abstractions.c#L716-L722

**Comment by momuno** [comment]
@rtheil-growlink Is the malloc failure only happening with either of these two callstacks that you have observed so far?
```
Error: Time:Wed Apr  6 07:58:10 2022 File:../lib/azure-iot-sdk-c/c-utility/pal/tlsio_options.c Func:tlsio_options_clone_option Line:119 unable to mallocAndStrcpy_s option value
Error: Time:Wed Apr  6 07:58:10 2022 File:../lib/azure-iot-sdk-c/c-utility/pal/tlsio_options.c Func:local_clone_option Line:231 Unexpected local_clone_option failure
Error: Time:Wed Apr  6 07:58:10 2022 File:../lib/azure-iot-sdk-c/c-utility/src/optionhandler.c Func:AddOptionInternal Line:76 unable to clone value
Error: Time:Wed Apr  6 07:58:10 2022 File:../lib/azure-iot-sdk-c/c-utility/src/optionhandler.c Func:OptionHandler_Clone Line:186 Error cloning option TrustedCerts
Error: Time:Wed Apr  6 07:58:10 2022 File:../lib/azure-iot-sdk-c/c-utility/src/optionhandler.c Func:AddOptionInternal Line:76 unable to clone value
Error: Time:Wed Apr  6 07:58:10 2022 File:../lib/azure-iot-sdk-c/c-utility/src/xio.c Func:xio_retrieveoptions Line:287 unable to OptionHandler_AddOption
```
```
Error: Time:Mon Apr 18 01:46:03 2022 File:../lib/azure-iot-sdk-c/c-utility/pal/tlsio_options.c Func:tlsio_options_clone_option Line:120 unable to mallocAndStrcpy_s option value: 12
Error: Time:Mon Apr 18 01:46:03 2022 File:../lib/azure-iot-sdk-c/c-utility/pal/tlsio_options.c Func:local_clone_option Line:232 Unexpected local_clone_option failure
Error: Time:Mon Apr 18 01:46:03 2022 File:../lib/azure-iot-sdk-c/c-utility/src/optionhandler.c Func:AddOptionInternal Line:76 unable to clone value
Error: Time:Mon Apr 18 01:46:03 2022 File:../lib/azure-iot-sdk-c/c-utility/pal/tlsio_options.c Func:tlsio_options_retrieve_options_ex Line:275 unable to save TrustedCerts option
Error: Time:Mon Apr 18 01:46:03 2022 File:../lib/azure-iot-sdk-c/c-utility/src/xio.c Func:xio_retrieveoptions Line:279 unable to concrete_io_retrieveoptions
```
I'm looking into the retrieveoptions logic to see if either there is a memory leak somewhere or see if anything changed here between 1.4 and 1.9

**Comment by rtheil-growlink** [comment]
Interesting. I didn't even notice that stack difference. This controller just did it again after I posted the log update, which is really weird. So I guess maybe the original reason I posted hasn't happened again, but now I got this new one.

I'm going to have to turn on my memory logging stuff and try to get the error again so I can see free and contiguous free, but in my previous tests there was more than enough contiguous space to malloc <622 bytes (max message size due to previous hardware platform limitations). Average message size is <150 bytes. Controller starts with more than 80k free on the heap once all is up and running and connected to Azure.

**Comment by momuno** [comment]
Thanks for the added info! Will update as I investigate further.

**Comment by rtheil-growlink** [comment]
After a bit of investigation with memory logging, this does seem to be a heap fragmentation issue. Something has changed and is causing contiguous memory to be reduced significantly. What I don't understand fully is that when the problem occurs, there's about 3.5k of contiguous heap, but the message being sent is <150 chars. Not sure how much it's trying to allocate at that point for such a small message.

**Comment by momuno** [comment]
@rtheil-growlink , after initial investigation, I see two things: 

1. There is a potential memory leak in `http_proxy_io.c` : `http_proxy_io_retrieve_options`, but this exists in both 1.4 and 1.9.  Are you using an HTTP proxy?
2. There was a memory leak in 1.4, fixed in 1.9, in `xio.c` : `xio_retrieveoptions`.  

However, both of these are contrary to your experience, having no memory allocation issue in 1.4 and instead seeing the issue in 1.9.

I am following up on the first issue with a separate PR.  Are you able to do any further testing comparing 1.4 and 1.9?  Please keep us posted if you are able to make any further progress.

**Comment by rtheil-growlink** [comment]
No HTTP proxies.

I have considered going back to 1.4 to see if there's a difference. One thing I have noticed today is that after the first SAS token expiration event, I lose about 4k of contiguous heap, but subsequent expirations do not reduce that. minHeapFree is contiguous heap:
```
W (2898528) memLog: after EvaluateSensors: LoopTaskFree: 47120/51200, HeapFree: 74516/195808, HeapDiff: 121292, minHeapFree: 38840
-> 20:25:54 DISCONNECT
W (2898608) AzureService: Connection Status result:IOTHUB_CLIENT_CONNECTION_UNAUTHENTICATED, Connection Status reason: IOTHUB_CLIENT_CONNECTION_EXPIRED_SAS_TOKEN
W (2898658) memLog: OneSecondLoop: LoopTaskFree: 47120/51200, HeapFree: 111660/195808, HeapDiff: 84148, minHeapFree: 36576
```

There's always a chance that the difference between 1.4 and 1.9 isn't the actual issue. There has been a significant amount of development on my part since the initial implementation, and even since the change to 1.9. I did discover what was dragging down the contiguous heap within our code, but it was only in certain situations so I will need to let this one run for a bit longer. However, since after the first SAS token expiration, I've not seen contiguous heap drop, so there's some good news.
